### DISCUSSION TITLE
file-diff and spellcheck modules not working with Textadept 12.2

### DISCUSSION THREAD

**Discussion by daron81** [orbitalquark/textadept#488] [Q&A]
Hello,

I recently upgrade to 12.2, and enabling either of these two modules produces an error.

For spellcheck, I get the following on startup:
`...modules/spellcheck/init.lua:262:bad argument #3 to 'newindex' (number expected, got nil)`

For file_diff, I get the following on startup:
`...modules/file_diff/init.lua:497:bad argument #3 to 'newindex' (number expected, got nil)`

Anyone else have this error besides me?

**Comment by orbitalquark** [comment]
Silly question, but did you also upgrade those modules to their latest versions? (i.e. from the modules.zip from the 12.2 release). If you did, are you using a custom theme that doesn't define a `view.colors` table? (e.g. https://github.com/orbitalquark/textadept/blob/3646cc46027556b43a9186311e5b6353a227c6c0/themes/light.lua#L15)

**Comment by daron81** [comment]
Using the default theme did indeed prevent the errors.

I am using v12.2 of the modules, and the theme here: https://github.com/rgieseke/base16-textadept/blob/main/base16/templates/textadept/themes/base16-outrun-dark.lua.

Is that table not initialized in the newer versions of TA?

**Reply by orbitalquark** [reply]
@rgieseke recently updated those Base16 themes: https://github.com/orbitalquark/textadept/discussions/476. Updating to one of those new themes should work for you.

**Reply by daron81** [reply]
I am using the newest versions of those themes, but still getting the error.

I am satisfied using the default theme for now, but I'd be curious to know if this error is the fault of the theme or not.

**Reply by rgieseke** [reply]
I can reproduce with 12.2, latest spellcheck and the updated base16 themes.

**Reply by rgieseke** [reply]
The base16 themes currently don't define a `colors.red`. Not sure whether it should define one or whether `spellcheck` should use an error style somehow?

**Reply by rgieseke** [reply]
As a workaround you could also add `colors.red` in your local copy of the base16 theme file I think.

```lua
colors.red = colors.base0A
```

**Comment by orbitalquark** [comment]
> The base16 themes currently don't define a `colors.red`. Not sure whether it should define one or whether `spellcheck` should use an error style somehow?

Oh, that's a good point. The file_diff module depends on `colors.red`, `colors.yellow`, and `colors.green`. I'm not sure how they would be changed to use a particular style since the colors will differ by theme.

Themes should not be responsible for knowing which colors modules may use, so at the very least there should be a more useful error about a particular color not being defined. That way users can add it to their theme manually if needed. I'll make a note to do this.

**Reply by rgieseke** [reply]
In the past I had red and green in the base16 templates: https://github.com/rgieseke/base16-textadept/commit/1a936ee87330dccb7b67ffecde0aa8eb762c83fa

Happy to add them again, but maybe there is a way for spellcheck to use the styles from the diff lexer?
Or we have other default styles that are a good fit.

Those I have included:

```lua
-- Diff.
styles.addition = {fore = colors.base0B}
styles.deletion = {fore = colors.base08}
styles.change = {fore = colors.base0E}
```

**Comment by orbitalquark** [comment]
I've added fields to the [spellcheck][1] and [diff][2] modules to override color names used:

```
local spellcheck = require('spellcheck')
spellcheck.misspelled_color_name = 'base08'

local file_diff = require('file_diff')
file_diff.addition_color_name = 'base0B'
file_diff.deletion_color_name = 'base08'
file_diff.modification_color_name = 'base0E'
```

The latest releases for each module will have this change. If you use them with custom themes and set the override color names, you should no longer get an error, and the colors should be correct.

[1]: https://github.com/orbitalquark/textadept-spellcheck/commit/268e0907b455ed2f7689c82cfa71b6893115d08a
[2]: https://github.com/orbitalquark/textadept-file-diff/commit/6537ca0aa40678b17a6cc7c51a0aff4c7c90c357

**Reply by rgieseke** [reply]
Thanks!
Couldn't there be a way to read style colors from e.g. 'error', 'number' or something else and then use those. There might be other themes which are not based on `base16` but might not define red or green either.

I think in Textredux something similar is done.

https://github.com/rgieseke/textredux/blob/main/core/style.lua

Here it uses the number style for example:

https://github.com/rgieseke/textredux/blob/92f5a844a55fb92278cd6cbf8dfad01c90f363f6/examples/buffer_styling.lua#L17

**Reply by orbitalquark** [reply]
You could use `view.styles[lexer.NUMBER].fore`, but as I hinted at earlier, if a theme colors numbers reddish, then showing a diff with deletions being red and additions being reddish would not look good. I think overrides are the way to go here, instead of assuming all themes color an entity the same-ish color.
